src/Chainweb/Miner/Coordinator.hs

```diff
@@ -20,71 +22,89 @@
 import Control.DeepSeq (NFData)
-import Control.Error.Util (hoistEither, (!?), (??))
-import Control.Lens (iforM, set, to, (^.), (^?!))
-import Control.Monad (join, unless)
-import Control.Monad.Trans.Class (lift)
-import Control.Monad.Trans.Except (runExceptT)
+import Control.Lens (view, (^?), (^?!))
+import Control.Monad
+import Control.Monad.Catch
 
+import Data.Aeson (ToJSON)
+import Data.Bool (bool)
 import qualified Data.ByteString as BS
 import Data.Foldable (foldl')
 import Data.Generics.Wrapped (_Unwrapped)
 import qualified Data.HashMap.Strict as HM
 import qualified Data.Map.Strict as M
+import qualified Data.Text as T
 import Data.Tuple.Strict (T2(..), T3(..))
 import qualified Data.Vector as V
 
 import GHC.Generics (Generic)
+import GHC.Stack
 
 import System.LogLevel (LogLevel(..))
 
 -- internal modules
 
 import Chainweb.BlockCreationTime
-import Chainweb.BlockHash (BlockHash, BlockHashRecord(..))
 import Chainweb.BlockHeader
-import Chainweb.BlockHeader.Validation (prop_block_pow)
-import Chainweb.BlockHeight
 import Chainweb.Cut hiding (join)
+import Chainweb.Cut.Create
 import Chainweb.Cut.CutHashes
 import Chainweb.CutDB
 import Chainweb.Logging.Miner
-import Chainweb.Miner.Pact (Miner(..), MinerId(..), minerId)
+import Chainweb.Miner.Pact (Miner(..), MinerId(..))
 import Chainweb.Payload
 import Chainweb.Sync.WebBlockHeaderStore
 import Chainweb.Time (Micros(..), Time(..), getCurrentTimeIntegral)
 import Chainweb.Utils hiding (check)
 import Chainweb.Version
+import Chainweb.Version.Utils
+import Chainweb.WebBlockHeaderDB
 
 import Data.LogMessage (JsonLog(..), LogFunction)
 
 import Utils.Logging.Trace (trace)
 
+-- -------------------------------------------------------------------------- --
+-- Utils
+
+-- | Lookup a 'BlockHeader' for a 'ChainId' in a cut and raise a meaningfull
+-- error if the lookup fails.
+--
+-- Generally, failing lookup in a cut is a code invariant violation. In almost
+-- all circumstances there should be a invariant in scope that guarantees that
+-- the lookup succeeds. This function is useful when debugging corner cases of
+-- new code logic, like graph changes.
+--
+lookupInCut :: HasCallStack => HasChainId cid => Cut -> cid -> BlockHeader
+lookupInCut c cid
+    | Just x <- c ^? ixg (_chainId cid) = x
+    | otherwise = error $ T.unpack
+        $ "Chainweb.Miner.Coordinator.lookupInCut: failed to lookup chain in cut."
+        <> " Chain: " <> sshow (_chainId cid) <> "."
+        <> " Cut Hashes: " <> encodeToText (cutToCutHashes Nothing c) <> "."
+
 -- -------------------------------------------------------------------------- --
 -- Miner
 
 -- | Precached payloads for Private Miners. This allows new work requests to be
 -- made as often as desired, without clogging the Pact queue.
 --
 newtype PrimedWork =
-    PrimedWork (HM.HashMap MinerId (HM.HashMap ChainId (Maybe PayloadWithOutputs)))
+    PrimedWork (HM.HashMap MinerId (HM.HashMap ChainId (Maybe PayloadData)))
```

Comment:
The outputs of a new block are not needed. All that is used is the payload hash and the `PayloadData`. The latter is attached to the new cut hashes before it is submitted to the cut pipeline.

Reply:
:+1: